src/Core/Services/MetadataProviders/SqlMetadataProvider.cs

```diff
@@ -325,15 +325,32 @@ private void LogPrimaryKeys()
             {
                 // row["DATA_TYPE"] has value type string so a direct cast to System.Type is not supported.
                 // See https://learn.microsoft.com/en-us/dotnet/framework/data/adonet/sql-server-data-type-mappings
-                Type systemType = SqlToCLRType((string)row["DATA_TYPE"]);
-                // Add to parameters dictionary without the leading @ sign
-                storedProcedureDefinition.Parameters.TryAdd(((string)row["PARAMETER_NAME"])[1..],
-                    new()
+                string sqlType = (string)row["DATA_TYPE"];
+                Type systemType = SqlToCLRType(sqlType);
+                ParameterDefinition paramDefinition = new()
+                {
+                    SystemType = systemType,
+                    DbType = TypeHelper.GetDbTypeFromSystemType(systemType)
+                };
+
+                if (GetDatabaseType() is DatabaseType.MSSQL &&
+                    (paramDefinition.SystemType == typeof(DateTime) || paramDefinition.SystemType == typeof(DateTimeOffset)))
+                {
+                    // MsSql types like date,smalldatetime,datetime,datetime2 are mapped to the same .NET type of DateTime.
+                    // Thus to determine the actual dbtype, we use the underlying MsSql type instead of the .NET type.
+                    DbType dbType;
+                    if (TryResolveDbType(sqlType, out dbType))
+                    {
+                        paramDefinition.DbType = dbType;
+                    }
+                    else
                     {
-                        SystemType = systemType,
-                        DbType = TypeHelper.GetDbTypeFromSystemType(systemType)
+                        _logger.LogWarning("Could not determine DbType for SqlDb type of {sqlType}", sqlType);
                     }
-                );
+                }
+
+                // Add to parameters dictionary without the leading @ sign
+                storedProcedureDefinition.Parameters.TryAdd(((string)row["PARAMETER_NAME"])[1..], paramDefinition);
```

Comment:
There should not be any duplicates here, so do we need to use `tryAdd`? If there is the possibility of duplicates and you want to avoid an exception here, right now this will ignore any attempts to add duplicates to the Stored Producedure Parameters, is that the intended behavior, that first should take precedent?

Reply:
I think, the use of TryAdd() is something that Ani has been suggesting to avoid any unforeseen exceptions. Duplicate stored procedure params, I don't think is something possible though but refraining to change this at this moment (Since this is not a change from this PR)